### PR TITLE
Issue/50 loss decay

### PR DESCRIPTION
**Pull request by peterdudfield**
Added a weighted exponentially decaying loss for MAE and MSE
This is calculated on every step. Thought this was a useful loss function to look at. 

Could be used  as the main loss function. 

Could also add other metrics here, but might not want to do all these every step, perhaps some just on validation step

### CODE REVIEW

**Review by jacobbieker — APPROVED** [review]
LGTM!

File: predict_pv_yield/models/metrics.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+def mse_each_forecast_horizon(output: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+    """
+    Get MSE for each forecast horizon
+    output - The model estimate of size (batch_size, forecast_length)
+    target - The truth of size (batch_size, forecast_length)
+
+    Return a tensor of size (forecast_length)
+    """
+
+    return torch.sum((output - target) ** 2, dim=0)
```
Comment (JackKelly) [inline predict_pv_yield/models/metrics.py:13]: Sorry if I'm being dumb but should this be `torch.mean` instead of `torch.sum`?  :)

(And same for `mae_each_forecast_horizon`)

If it is meant to be `sum` instead of `mean` (because you weight the values elsewhere) then maybe the function should have a different name so people don't use this function thinking it computes the mean!?  e.g. `sum_abs_error_...`???

File: predict_pv_yield/models/loss.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import math
+import torch
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class WeightedLosses:
+    def __init__(self, decay_rate: int = None, forecast_length: int = 6):
```
Comment (JackKelly) [inline predict_pv_yield/models/loss.py:9]: Given that `decay_rate` can be `None`, I'd recommend using this type annotation:  

```python
decay_rate: Optional[int] = None
```

Just to make it super-explicit that `decay_rate` can be `None`!  :)

File: predict_pv_yield/models/metrics.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import torch
+
+
+def mse_each_forecast_horizon(output: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
```
Comment (JackKelly) [inline predict_pv_yield/models/metrics.py:4]: Love these short functions to aid readability of the calling code!

File: tests/models/test_loss.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+def test_mae_exp():
+    forecast_length = 2
+    w = WeightedLosses(forecast_length=forecast_length)
+
+    output = torch.Tensor([1, 3])
+    target = torch.Tensor([1, 5])
+
+    loss = w.get_mae_exp(output=output, target=target)
+
+    assert loss == pytest.approx(2 / 3)  # (1-1)*2/3 + (5-3)*1/4
```
Comment (JackKelly) [inline tests/models/test_loss.py:23]: I like that you 'show your working' in the comment :)

**Review by JackKelly — COMMENTED** [review]
Great work!  Thank you!  Just a few tiny comments... feel free to ignore!...

File: predict_pv_yield/models/metrics.py (hunk 1 above)
Comment (peterdudfield) [reply predict_pv_yield/models/metrics.py:13]: Yea, you are totally right, it should have bee torch.mean

File: predict_pv_yield/models/loss.py (hunk 2 above)
Comment (peterdudfield) [reply predict_pv_yield/models/loss.py:9]: Yea, I like it

File: predict_pv_yield/models/metrics.py (hunk 3 above)
Comment (peterdudfield) [reply predict_pv_yield/models/metrics.py:4]: thanks

File: tests/models/test_loss.py (hunk 4 above)
Comment (peterdudfield) [reply tests/models/test_loss.py:23]: thanks
